src/bsdfs/bumpmap.cpp

```diff
@@ -0,0 +1,210 @@
+   :caption: Roughplastic BSDF with bump mapping
+.. subfigend::
+   :label: fig-bumpmap
+
+The following XML snippet describes a rough plastic material affected by a bump
+map. Note the we set the ``raw`` properties of the bump map ``bitmap`` object to
+``true`` in order to disable the transformation from sRGB to linear encoding:
+
+.. code-block:: xml
+    :name: bumpmap
+
+    <bsdf type="bumpmap">
+        <texture name="arbitrary" type="bitmap">
+            <boolean name="raw" value="true"/>
+            <string name="filename" value="textures/bumpmap.jpg"/>
+        </texture>
+        <bsdf type="roughplastic"/>
+    </bsdf>
+
+*/
+
+template <typename Float, typename Spectrum>
+class BumpMap final : public BSDF<Float, Spectrum> {
+public:
+    MTS_IMPORT_BASE(BSDF, m_flags, m_components)
+    MTS_IMPORT_TYPES(Texture)
+
+    BumpMap(const Properties &props) : Base(props) {
+        for (auto &[name, obj] : props.objects(false)) {
+            auto bsdf = dynamic_cast<Base *>(obj.get());
+            if (bsdf) {
+                if (m_nested_bsdf)
+                    Throw("Only a single BSDF child object can be specified.");
+                m_nested_bsdf = bsdf;
+                props.mark_queried(name);
+            }
+            auto texture = dynamic_cast<Texture *>(obj.get());
+            if (texture) {
+                if (m_nested_texture)
+                    Throw("Only a single Texture child object can be specified.");
+                m_nested_texture = texture;
+                props.mark_queried(name);
+            }
+        }
+        if (!m_nested_bsdf)
+            Throw("Exactly one BSDF child object must be specified.");
+        if (!m_nested_texture)
+            Throw("Exactly one Texture child object must be specified.");
+
+        m_scale = props.float_("scale", 1.f);
+
+        // Add all nested components
+        m_components.clear();
+        for (size_t i = 0; i < m_nested_bsdf->component_count(); ++i)
+            m_components.push_back(m_nested_bsdf->flags(i));
+        m_flags = m_nested_bsdf->flags();
+    }
+
+    std::pair<BSDFSample3f, Spectrum> sample(const BSDFContext &ctx,
+                                             const SurfaceInteraction3f &si,
+                                             Float sample1,
+                                             const Point2f &sample2,
+                                             Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::BSDFSample, active);
+
+        // Sample nested BSDF with perturbed shading frame
+        SurfaceInteraction3f perturbed_si(si);
+        perturbed_si.sh_frame = frame(si, active);
+        perturbed_si.wi = perturbed_si.to_local(si.wi);
+        auto [bs, weight] = m_nested_bsdf->sample(ctx, perturbed_si,
+                                                  sample1, sample2, active);
+        active &= any(neq(depolarize(weight), 0.f));
```

Comment:
A minor comment here:  the `none(active)` call below will trigger an expensive horizontal operation to the GPU that flushes all computation and waits for an answer whether any of the entries is still alive. (So a kernel compilation, flush, and PCI express read that stalls for the GPU).

Because the answer is almost inevitably: -- yes, there is still one lane active --, we generally use `none_or<true>` which will use the default template argument on in the JIT-compiled GPU mode and perform the horizontal operation on all other targets.

In this case, you are only saving a few multiplications, so it's probably not worth it to have this extra check at all. I would just remove it entirely.